FWCore/Framework/interface/Callback.h

```diff
@@ -49,31 +60,249 @@ namespace edm {
+              ServiceRegistry::Operate guard(serviceToken_.lock());
+              record_->activityRegistry()->postESModulePrefetchingSignal_.emit(record_->key(),
+                                                                               callback_->callingContext_);
+            });
+          } catch (cms::Exception& iException) {
+            if (not excptr) {
+              edm::exceptionContext(iException, callback_->callingContext_);
+              excptr = std::current_exception();
+            }
+          }
+          if (excptr) {
+            holder_.doneWaiting(excptr);
+            return;
+          }
+
+          auto group = holder_.group();
+          callback_->producer_->queue().push(
+              *group,
+              [callback = callback_,
+               holder = std::move(holder_),
+               serviceToken = std::move(serviceToken_),
+               record = record_,
+               eventSetupImpl = eventSetupImpl_]() mutable {
+                callback->callingContext_.setState(ESModuleCallingContext::State::kRunning);
+                std::exception_ptr exceptPtr;
+                try {
+                  convertException::wrap([&callback, &holder, &serviceToken, &record, &eventSetupImpl] {
+                    ESModuleCallingContext const& context = callback->callingContext_;
+                    auto proxies = callback->getTokenIndices();
+                    if (callback->postMayGetProxies_) {
+                      proxies = &((*callback->postMayGetProxies_).front());
+                    }
+                    TRecord rec;
+                    edm::ESParentContext pc{&context};
+                    rec.setImpl(record, callback->transitionID(), proxies, eventSetupImpl, &pc);
+                    ServiceRegistry::Operate operate(serviceToken.lock());
+                    record->activityRegistry()->preESModuleAcquireSignal_.emit(record->key(), context);
+                    struct EndGuard {
+                      EndGuard(EventSetupRecordImpl const* iRecord, ESModuleCallingContext const& iContext)
+                          : record_{iRecord}, context_{iContext} {}
+                      ~EndGuard() {
+                        record_->activityRegistry()->postESModuleAcquireSignal_.emit(record_->key(), context_);
+                      }
+                      EventSetupRecordImpl const* record_;
+                      ESModuleCallingContext const& context_;
+                    };
+                    EndGuard guard(record, context);
+                    (*callback->acquireFunction_)(rec, holder);
+                  });
+                } catch (cms::Exception& iException) {
+                  iException.addContext("Running acquire");
+                  exceptPtr = std::current_exception();
+                }
+                holder.doneWaiting(exceptPtr);
+              });
+        }
+
+      private:
+        Callback* callback_;
+        WaitingTaskWithArenaHolder holder_;
```

Comment:
Maybe I'll express myself better with an example. E.g. in `Worker.h` we currently have
```cpp
void Worker::runAcquireAfterAsyncPrefetch(..., WaitingTaskWithArenaHolder holder) {
  ...
  runAcquire(.., holder);
  ...
  holder.doneWaiting(...);
}
```
where the `WaitingTaskWithArenaHolder::doneWaiting()` may call `task_arena::enqueue()`.

I'm wondering if changing it to
```cpp
void Worker::runAcquireAfterAsyncPrefetch(..., WaitingTaskHolder holder) {
  ...
  runAcquire(.., WaitingTaskWithArenaHolder(holder));
  ...
  holder.doneWaiting(...);
}
```
would make sense. The `WaitingTaskWithArenaHolder` constructor would still be called by the framework, and therefore it should pick the same `task_arena`, but I believe in this way if/when the task held by the holder gets "launched" by the `doneWaiting()` called by the framework, the task would be run without the `task_arena::enqueue()`.

Reply:
Ahhh, now I understand what you are suggesting.

There is a constructor:
explicit WaitingTaskWithArenaHolder(WaitingTaskHolder&& iTask);

but I don't see
explicit WaitingTaskWithArenaHolder(WaitingTaskHolder const& iTask);

Is there a reason for that Chris @Dr15Jones? I suppose I could try to
implement that.

I could pass around a WaitingTask* and make both the holders from that.
I could do this. My only argument against this is that the probability
of the ExternalWork completing that fast is small. On the other hand,
it isn't hard to do. If this unlikely situation occurred, it would improve
performance.

Reply:
> I suppose I could try to implement that.

I would in any case leave this possibility beyond this PR, so that it could be done consistently in the Event and EventSetup sides (which certainly is beyond this PR).

> My only argument against this is that the probability of the ExternalWork completing that fast is small.

That may be, on the other hand we don't really know right now. With EventSetup I came to consider this question, because the majority (if not all) use cases in the near term are transfers of data structures from host to GPU, and some of those data structures may be small. This is probably another argument for leaving the possible change later, so that we could test its impact on actual application.

Reply:
Sounds good. I will make no changes in this PR and leave this for the future.

I'm still looking at your other comments. Those require some thought.